File: jscomp/test/splice_test.js
'use strict';

var Mt = require("./mt.js");
var Caml_array = require("../../lib/js/caml_array.js");
var Caml_splice_call = require("../../lib/js/caml_splice_call.js");

var suites = {
  contents: /* [] */0
};

var test_id = {
  contents: 0
};

function eq(loc, x, y) {
  Mt.eq_suites(test_id, suites, loc, x, y);
}

var Caml_splice_call$1 = {};

Math.max(1);

function f00(a, b) {
  return a.send(b);
}

var a = [];

a.push(1, 2, 3, 4);

eq("File \"splice_test.ml\", line 29, characters 5-12", a, [
      1,
      2,
      3,
      4
    ]);

function dynamic(arr) {
  var a = [];
  Caml_splice_call.spliceObjApply(a, "push", [
        1,
        arr
      ]);
  eq("File \"splice_test.ml\", line 34, characters 5-12", a, Caml_array.concat({
            hd: [1],
            tl: {
              hd: arr,
              tl: /* [] */0
            }
          }));
}

dynamic([
      2,
      3,
      4
    ]);

dynamic([]);

dynamic([
      1,
      1,
      3
    ]);

var a$1 = [];

a$1.push(1, 2, 3, 4);

eq("File \"splice_test.ml\", line 51, characters 7-14", a$1, [
      1,
      2,
      3,
      4
    ]);

function dynamic$1(arr) {
  var a = [];
  Caml_splice_call.spliceObjApply(a, "push", [
        1,
        arr
      ]);
  eq("File \"splice_test.ml\", line 56, characters 7-14", a, Caml_array.concat({
            hd: [1],
            tl: {
              hd: arr,
              tl: /* [] */0
            }
          }));
}

dynamic$1([
      2,
      3,
      4
    ]);

dynamic$1([]);

dynamic$1([
      1,
      1,
      3
    ]);

var Pipe = {
  dynamic: dynamic$1
};

function f1(c) {
  return Caml_splice_call.spliceApply(Math.max, [
              1,
              c
            ]);
}

eq("File \"splice_test.ml\", line 67, characters 6-13", Math.max(1, 2, 3), 3);

eq("File \"splice_test.ml\", line 68, characters 6-13", Math.max(1), 1);

eq("File \"splice_test.ml\", line 69, characters 6-13", Math.max(1, 1, 2, 3, 4, 5, 2, 3), 5);

Mt.from_pair_suites("splice_test.ml", suites.contents);

exports.suites = suites;
exports.test_id = test_id;
exports.eq = eq;
exports.Caml_splice_call = Caml_splice_call$1;
exports.f00 = f00;
exports.dynamic = dynamic;
exports.Pipe = Pipe;
exports.f1 = f1;
/*  Not a pure module */

File: jscomp/test/splice_test.ml
let suites :  Mt.pair_suites ref  = ref []
let test_id = ref 0
let eq loc x y = Mt.eq_suites ~test_id ~suites loc x y


module Caml_splice_call = struct
end
external f : int -> int array -> int = "Math.max"
  [@@mel.variadic]

;; f 1 [||] |. ignore

external send : int -> int array -> int = "send"
  [@@mel.variadic] [@@mel.send]

let f00 a b =
  a |. send [|b|]

external push : int array -> int -> int array -> unit =
  "push" [@@mel.send] [@@mel.variadic]

(* This is only test, the binding maybe wrong
  since in OCaml array'length is not mutable
*)
let () =
  let a = [||] in
  a |. push 1 [|2;3;4|];

  eq __LOC__ a [|1;2;3;4|]

let dynamic arr =
  let a = [||] in
  a |. push 1 arr ;
  eq __LOC__ a (Array.concat [[|1|]; arr])

;; dynamic [|2;3;4|]
;; dynamic [||]
;; dynamic [|1;1;3|]

(* Array constructor with a single parameter `x`
   just makes an array with its length set to `x`,
   so at least two parameters are needed
*)
external newArr : int -> int -> int array -> int array = "Array"
  [@@mel.variadic] [@@mel.new]

let () =
  let a = newArr 1 2 [|3;4|] in
  eq __LOC__ a [|1;2;3;4|]

let dynamicNew arr =
  let a = newArr 1 2 arr in
  eq __LOC__ a (Array.concat [[|1; 2|]; arr])

;; dynamicNew [|3;4|]
;; dynamicNew [||]
;; dynamicNew [|1;3|]

module Pipe = struct
  external push :  int -> int array -> (int array [@mel.this]) -> unit =
    "push" [@@mel.send] [@@mel.variadic]

  (* This is only test, the binding maybe wrong
     since in OCaml array'length is not mutable
  *)
  let () =
    let a = [||] in
    a |> push 1 [|2;3;4|];

    eq __LOC__ a [|1;2;3;4|]

  let dynamic arr =
    let a = [||] in
    a |> push 1 arr ;
    eq __LOC__ a (Array.concat [[|1|]; arr])

  ;; dynamic [|2;3;4|]
  ;; dynamic [||]
  ;; dynamic [|1;1;3|]

end

let f1 (c : int array) =  f 1 c

;; eq __LOC__ (f1  [|2;3|] ) 3
;; eq __LOC__ (f1  [||] ) 1
;; eq __LOC__ (f1 [|1;2;3;4;5;2;3|]) 5

;; Mt.from_pair_suites __FILE__ !suites
